### ISSUE TITLE
`Literal` should not be assignable to `type[T]`

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#6895]
Pyright should emit an error in the following case but does not.

```python
def func1(x: type[Any]):
    ...

func1(Literal[0])
```

**Comment by hmc-cs-mdrissi** [comment]
Hmm I don't follow why given that currently type allows Union/other special forms for pyright. What makes Union/Annotated fine, but Literal not?

I realize that Union/special forms being allowed is also discrepancy with mypy but until [typeform](https://github.com/python/mypy/issues/9773) exists, I've found it pretty helpful that,

```python
def load_config(ty: type[T], config: object) -> T:
  ...

load_config(Union[str, int], 5)
```

works.

**Comment by erictraut** [comment]
The type `type[T]` means "a subtype of the `type` class" — that is, an instantiable class object. `Literal` does not represent an instantiable class, so I don't think it should be compatible with `type[T]`. By contrast, the type `str` and `int` and `str | int` all represent subtypes of `type`.

**Comment by Viicos** [comment]
For reference, `mypy` doesn't handle `Union` correctly, i.e.:

```python
def load_config(ty: type[T], config: object) -> T:
  ...

load_config(Union[str, int], 5)  # Argument 1 to "load_config" has incompatible type "object"; expected "type[Never]"
```

By switching from `type[T]` to `T`:

```python
load_config(Union[str, int], 5)  # Revealed type is "builtins.object". No error but still not what would be expected
```

I'm wondering if this should be added to the spec or should this wait for the typeform proposal?

**Comment by erictraut** [comment]
This is included in pyright 1.1.346, which I just published. It will also be included in a future release of pylance.

**Comment by adampauls** [comment]
I see that 
```python
Input = TypeVar("Input")

class Foo(Generic[Input]):
    def __init__(
        self,
        input_type: type[Input],
    ) -> None:
        pass


Foo(int | str) #  Argument of type "type[int] | type[str]" cannot be assigned to parameter "input_type" of type "type[Input@Foo]"
```

Is now an error.  But you say [here](https://github.com/microsoft/pyright/issues/6895#issuecomment-1876039789) that 
> str and int and str | int all represent subtypes of type.

What is the right type now for a variable that can be a union type?

**Comment by erictraut** [comment]
If you would like to pass one or more classes as an argument, you could use a collection class like a tuple (or set or list).

```python
from typing import Generic, TypeVar

Input = TypeVar("Input")

class Foo(Generic[Input]):
    def __init__(self, input_types: tuple[type[Input], ...]) -> None:
        pass

f = Foo((int, str))
reveal_type(f) # Foo[int, str]
```

There has been some discussion of introducing a `TypeForm` symbol that could be used in annotations, but the details have yet to be worked out for such a construct.

**Comment by adampauls** [comment]
There's still something I'm missing. `type(int|str)` is `types.UnionType`, so it seems like I should be able to pass both `int` and `int | str` to a field of type `type[Any] | UnionType`, but

```python
from types import UnionType
from typing import Any, TypeVar

Input = TypeVar("Input")


def foo1(type_: type[Any] | type[UnionType]) -> None:
    pass


def foo2(type_: type[Any]) -> None:
    pass


def foo3(type_: type[Any] | UnionType) -> None:
    pass


#   error: Argument of type "type[int] | type[str]" cannot be assigned to parameter "type_" of type "type[Any] | type[UnionType]" in function "foo1"
#     Type "type[UnionType]" cannot be assigned to type "type[Any] | type[UnionType]"
#       Type "type[UnionType]" cannot be assigned to type "type[UnionType]" (reportArgumentType)
foo1(int | str)

# error: Argument of type "type[int] | type[str]" cannot be assigned to parameter "type_" of type "type[Any]" in function "foo2" (reportArgumentType)
foo2(int | str)

#  error: Argument of type "type[int] | type[str]" cannot be assigned to parameter "type_" of type "type[Any] | UnionType" in function "foo3"
#    Type "type[UnionType]" cannot be assigned to type "type[Any] | UnionType"
#      Type "type[type]" is incompatible with "type[UnionType]" (reportArgumentType)
foo3(int | str)
```

` Type "type[UnionType]" cannot be assigned to type "type[UnionType]" (reportArgumentType)` from `foo1` seems like something has gone wrong, as does the appearance of `type[UnionType]` in `foo3`.

This is all on 1.1.348. Should I file an issue, or am I misunderstanding something?

**Comment by erictraut** [comment]
Please file a bug. I agree that the call to `foo3` should work. Pyright is currently converting `int | str` to `type[UnionType]`, when it's used as a runtime value, but it should be converting it to `UnionType`.

**Comment by adampauls** [comment]
Done [here](https://github.com/microsoft/pyright/issues/7110).
